### PR TITLE
Adding support for setting the device id at runtime.

### PR DESCRIPTION
**Pull request by pschuh**
The device id is set per-thread. There is still work to be done in order to make this seamlessly work with withDevice() and multiple threads. The state should be essentially cloned when a new thread is started.

### CODE REVIEW

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Review by mhong — APPROVED** [review]
Very nice!

Once the tests pass, it'd be great to submit this PR, so that I can rebase https://github.com/apple/swift/pull/21406 on top of this.

Please feel free to defer addressing comments (unless there are significant concerns with the approach taken here).

**Comment by mhong** [comment]
@rxwei , would be great to get your feedback on swift code as usual.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 1]
```diff
@@ -1548,63 +1544,97 @@ func _TFCOpSetAttrStringArray(_ op: CTFEOp,
...
 
-internal extension _ExecutionContext {
-  var currentDeviceKind: DeviceKind {
-    return sync {
-      deviceScopes.last ?? .default
+@usableFromInline
+class _ThreadLocalState {
+  var deviceScopes: [(DeviceKind, UInt)?] = []
+
+  private static let key = { () -> pthread_key_t in
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: - Adding explicit types to non-trivial static variable declarations reduces type inference complexity.
- Not specifying closure types is generally preferred.

```suggestion
  private static let key: pthread_key_t = {
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 2]
```diff
@@ -1548,63 +1544,97 @@ func _TFCOpSetAttrStringArray(_ op: CTFEOp,
...
   // TODO: TPU?
 }
 
-internal extension _ExecutionContext {
-  var currentDeviceKind: DeviceKind {
-    return sync {
-      deviceScopes.last ?? .default
+@usableFromInline
+class _ThreadLocalState {
+  var deviceScopes: [(DeviceKind, UInt)?] = []
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]:
```suggestion
  var deviceScopes: [(kind: DeviceKind, index: UInt)?] = []
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 3]
```diff
@@ -1548,63 +1544,97 @@ func _TFCOpSetAttrStringArray(_ op: CTFEOp,
...
+/// device.
 ///
 /// - Parameters:
 ///   - kind: A kind of device to run TensorFlow operations on.
+///   - index: The device to run the ops on.
+///   - body: A closure whose TensorFlow operations are to be executed on the
+///     specified kind of device.
+@inlinable
+public func withDevice<R>(_ kind: DeviceKind, _ index: UInt = 0,
+                       perform body: () throws -> R) rethrows -> R {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: Align args.

File: stdlib/public/TensorFlow/Threading.swift [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    )
+    internalConsistencyCheck(creationStatus == 0)
+    self.thread = newThread
+  }
+  func join() {
+    internalConsistencyCheck(pthread_join(thread, nil) == 0)
+  }
+}
+
+public func runOnNDevices(_ n: Int, perform body: @escaping (Int) -> ()) {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Threading.swift]: Why is this public?

Comment (pschuh) [reply stdlib/public/TensorFlow/Threading.swift]: Needs to be for the unit test for collective ops. We're also going to need something like this (maybe also with an implicit withDevice(.device_type, i) {} for when we do parallel training. Mingsheng and I are still working through the details.

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/Threading.swift (hunk 4 above)
Comment (rxwei) [reply stdlib/public/TensorFlow/Threading.swift]: I see. It is pretty far from being a Swifty user API, but defining it here is fine for testing. In these cases, we should at least make the function name start with an underscore.

**Review by rxwei — APPROVED** [review]

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Pull request merged**